////////////////////////////////////////////////////////////////////////////////
// Copyright (c) 2014-2019, Lawrence Livermore National Security, LLC.
// Produced at the Lawrence Livermore National Laboratory.
// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
//
// LLNL-CODE-697807.
// All rights reserved.
//
// This file is part of LBANN: Livermore Big Artificial Neural Network
// Toolkit. For details, see http://software.llnl.gov/LBANN or
// https://github.com/LLNL/LBANN.
//
// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
// may not use this file except in compliance with the License.  You may
// obtain a copy of the License at:
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied. See the License for the specific language governing
// permissions and limitations under the license.
////////////////////////////////////////////////////////////////////////////////

syntax = "proto3";

import "optimizers.proto";
import "layers.proto";

package lbann_data;

message Weights {
  string name = 1;
  Optimizer optimizer = 2;
  Initializer initializer = 3;
  DataType datatype = 4;
}

message Initializer {
  oneof initializer_type {
    ConstantInitializer constant_initializer = 20;
    ValueInitializer value_initializer = 21;
    UniformInitializer uniform_initializer = 22;
    NormalInitializer normal_initializer = 23;
    GlorotNormalInitializer glorot_normal_initializer = 24;
    GlorotUniformInitializer glorot_uniform_initializer = 25;
    HeNormalInitializer he_normal_initializer = 26;
    HeUniformInitializer he_uniform_initializer = 27;
    LeCunNormalInitializer lecun_normal_initializer = 28;
    LeCunUniformInitializer lecun_uniform_initializer = 29;
  }

  // Weight initializers
  message ConstantInitializer {
    double value = 1;
  }
  message ValueInitializer {
    string values = 1;
  }
  message UniformInitializer {
    double min = 1;
    double max = 2;
  }
  message NormalInitializer {
    double mean = 1;
    double standard_deviation = 2;
  }
  message GlorotNormalInitializer {}
  message GlorotUniformInitializer {}
  message HeNormalInitializer {}
  message HeUniformInitializer {}
  message LeCunNormalInitializer {}
  message LeCunUniformInitializer {}
}
